lens-cube/src/main/java/org/apache/lens/cube/parse/StorageCandidate.java

```diff
@@ -139,9 +177,310 @@ public boolean contains(Candidate candidate) {
+    partColsQueried.add(timeRange.getPartitionColumn());
+    StringBuilder extraWhereClause = new StringBuilder();
+    Set<FactPartition> rangeParts = getPartitions(timeRange, validUpdatePeriods, true, failOnPartialData, missingParts);
+    String partCol = timeRange.getPartitionColumn();
+    boolean partColNotSupported = rangeParts.isEmpty();
+    String storageTableName = getStorageName();
+    partColNotSupported &=
+      skipStorageCauses.containsKey(storageTableName) && skipStorageCauses.get(storageTableName).getCause()
+        .equals(PART_COL_DOES_NOT_EXIST) && skipStorageCauses.get(storageTableName).getNonExistantPartCols()
+        .contains(partCol);
+    TimeRange prevRange = timeRange;
+    String sep = "";
+    while (rangeParts.isEmpty()) {
+      String timeDim = cubeql.getBaseCube().getTimeDimOfPartitionColumn(partCol);
+      if (partColNotSupported && !getFact().getColumns().contains(timeDim)) {
+        unsupportedTimeDims.add(cubeql.getBaseCube().getTimeDimOfPartitionColumn(timeRange.getPartitionColumn()));
+        break;
+      }
+      TimeRange fallBackRange = getFallbackRange(prevRange, this.getFact().getName(), cubeql);
+      log.info("No partitions for range:{}. fallback range: {}", timeRange, fallBackRange);
+      if (fallBackRange == null) {
+        break;
+      }
+      partColsQueried.add(fallBackRange.getPartitionColumn());
+      rangeParts = getPartitions(fallBackRange, validUpdatePeriods, true, failOnPartialData, missingParts);
+      extraWhereClause.append(sep)
+        .append(prevRange.toTimeDimWhereClause(cubeql.getAliasForTableName(cubeql.getCube()), timeDim));
+      sep = " AND ";
+      prevRange = fallBackRange;
+      partCol = prevRange.getPartitionColumn();
+      if (!rangeParts.isEmpty()) {
+        break;
+      }
+    }
+    if (!unsupportedTimeDims.isEmpty()) {
+      log.info("Not considering fact table:{} as it doesn't support time dimensions: {}", this.getFact(),
+        unsupportedTimeDims);
+      cubeql.addFactPruningMsgs(this.getFact(), timeDimNotSupported(unsupportedTimeDims));
+      return false;
+    }
+    Set<String> nonExistingParts = missingParts.toSet(partColsQueried);
+    nonExistingPartitions.addAll(nonExistingParts);
+    if (rangeParts.size() == 0 || (failOnPartialData && !nonExistingParts.isEmpty())) {
+      log.info("No partitions for fallback range:{}", timeRange);
+      return false;
+    }
+    String extraWhere = extraWhereClause.toString();
+    if (!StringUtils.isEmpty(extraWhere)) {
+      rangeToWhere.put(storageTableName, "((" + rangeWriter
+        .getTimeRangeWhereClause(cubeql, cubeql.getAliasForTableName(cubeql.getCube().getName()), rangeParts)
+        + ") and  (" + extraWhere + "))");
+    } else {
+      rangeToWhere.put(storageTableName, rangeWriter
+        .getTimeRangeWhereClause(cubeql, cubeql.getAliasForTableName(cubeql.getCube().getName()), rangeParts));
+    }
+    this.storagePartitions = rangeParts;
```

Comment:
Can we directly assign ? Should we not check if it already has some partitions, in that case we should add the extra ones.